common/persistence/cassandra/queue_v2_store.go

```diff
@@ -0,0 +1,242 @@
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/nosql/nosqlplugin/cassandra/gocql"
+	"golang.org/x/exp/slices"
+)
+
+type (
+	// queueV2Store contains the SQL queries and serialization/deserialization functions to interact with the queues and
+	// queue_messages tables that implement the QueueV2 interface. The schema is located at:
+	//	schema/cassandra/temporal/versioned/v1.9/queues.cql
+	queueV2Store struct {
+		session gocql.Session
+	}
+)
+
+const (
+	TemplateEnqueueMessageQuery  = `INSERT INTO queue_messages (queue_type, queue_name, queue_partition, message_id, message_payload, message_encoding) VALUES (?, ?, ?, ?, ?, ?) IF NOT EXISTS`
+	TemplateGetMessagesQuery     = `SELECT message_id, message_payload, message_encoding FROM queue_messages WHERE queue_type = ? AND queue_name = ? AND queue_partition = ? AND message_id >= ? ORDER BY message_id ASC LIMIT ?`
+	TemplateGetMaxMessageIDQuery = `SELECT MAX(message_id) AS message_id FROM queue_messages WHERE queue_type = ? AND queue_name = ? AND queue_partition = ?`
+)
+
+var (
+	// ErrQueueMessageConflict occurs when a message with the same ID already exists in the queue.
+	// This is possible when there are concurrent writes to the queue because we enqueue a message using two queries:
+	//
+	// 	1. SELECT MAX(ID) to get the next message ID (for a given queue partition)
+	// 	2. INSERT (ID, message) with CAS
+	//
+	// See the following example:
+	//
+	//  Client A           Client B                  Cassandra DB
+	//  |                  |                                    |
+	//  |---1. SELECT MAX(ID)---------------------------------->|
+	//  |                  |                                    |
+	//  |<--2. Return ID X--------------------------------------|
+	//  |                  |                                    |
+	//  |                  |---3. SELECT MAX(ID)--------------->|
+	//  |                  |                                    |
+	//  |                  |<--4. Return ID X-------------------|
+	//  |                  |                                    |
+	//  |---5. INSERT (X+1, msgA) with CAS--------------------->|
+	//  |                  |                                    |
+	//  |<--6. Acknowledge--------------------------------------|
+	//  |                  |                                    |
+	//  |                  |---7. INSERT (X+1, msgB) with CAS-->|
+	//  |                  |                                    |
+	//  |                  |<--8. Conflict/Error----------------|
+	//  |                  |                                    |
+	ErrQueueMessageConflict            = errors.New("queue message with id already exists, likely due to concurrent writes")
+	ErrInvalidQueueMessageEncodingType = errors.New("invalid encoding type for queue message")
+)
+
+func NewQueueV2Store(session gocql.Session) persistence.QueueV2 {
+	return &queueV2Store{
+		session: session,
+	}
+}
+
+func (q *queueV2Store) EnqueueMessage(
```

Comment:
we need to control the concurrency of this method. plz add a TODO comment.

Reply:
Added a TODO